plutus-metatheory/src/VerifiedCompilation/UCaseOfCase.lagda.md

```diff
@@ -0,0 +1,100 @@
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip alts alts') -- recursive translation for the other case patterns 
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip tt tt') -- recursive translation for the true branch
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip ft ft') -- recursive translation for the false branch
+                ----------------------
+                → X ⊢̂
+                       (case ((((force (builtin ifThenElse)) · b) · (constr tn tt)) · (constr fn ft)) alts)
+                       ⊳̂ ((((force (builtin ifThenElse)) · b) · (case (constr tn' tt') alts')) · (case (constr tn' tt') alts'))
+   var : ∀ {x : X} → X ⊢̂ (` x) ⊳̂ (` x) 
+   ƛ   : ∀ {x x' : Maybe X ⊢}
+           → Maybe X ⊢̂ x ⊳̂ x'
+           ----------------------
+           →  X ⊢̂ ƛ x ⊳̂ ƛ x' 
+   _·_ : ∀ {f t f' t' : X ⊢} → (X ⊢̂ f ⊳̂ f') → (X ⊢̂ t ⊳̂ t') → (X ⊢̂ (f · t) ⊳̂ (f' · t'))
+   force : ∀ {t t' : X ⊢} → X ⊢̂ t ⊳̂ t' → X ⊢̂ force t ⊳̂ force t'  
+   delay : ∀ {t t' : X ⊢} → X ⊢̂ t ⊳̂ t' → X ⊢̂ delay t ⊳̂ delay t'  
+   con : ∀ {tc : TmCon} → X ⊢̂ con tc ⊳̂ con tc
+   constr : ∀ {xs xs' : List (X ⊢)} { n : ℕ }
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip xs xs')
+                ------------------------
+                → X ⊢̂ constr n xs ⊳̂ constr n xs' 
+   case :  ∀ {p p' : X ⊢} {alts alts' : List (X ⊢)}
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip alts alts') -- recursive translation for the other case patterns
+                → X ⊢̂ p ⊳̂ p'
+                ------------------------
+                → X ⊢̂ case p alts ⊳̂ case p alts' 
+   builtin : ∀ {b : Builtin} → X ⊢̂ builtin b ⊳̂ builtin b
+   error : X ⊢̂ error ⊳̂ error
+
+```
+
+## Decision Procedure
+
+Since this compiler phase is just a syntax re-arrangement in a very particular situation, we can
+match on that situation in the before and after ASTs and apply the translation rule for this, or
+expect anything else to be unaltered.
+
+This can just traverse the ASTs applying the constructors from the transition relation.
+
+```
+--_⊢̂_⊳̂?_ : {X : Set} (ast ast' : X ⊢) → Dec (X ⊢̂ ast ⊳̂ ast')
+--_⊢̂_⊳̂?_ ast ast' = ?
+
+```
+
+## Semantic Equivalence
+
+We can show that this translation never alters the semantics of the statement. This is shown
+in terms of the CEK machine evaluation. Since it is a simple re-arrangement of the syntax, it
+isn't a refinement argument - the state before and after the operation is the same type, and is
+unaltered buy the syntax re-arrangement.
+
+This does rely on the encoding of the semantics of `IfThenElse` in the CEK module, since we
+need to show that the effective list of cases is the same as it would have been without the re-arrangement.
+
+The `stepper` function uses the CEK machine to evaluate a term. Here we call it with a very
```

Comment:
I don't mind using the CEK machine, but why? Why not simply small-step/big-step semantics? And then test that those match with the behavior of the CEK machine if we want to ensure that this is the case.

Reply:
I was led to believe the CEK machine was closer to what "really happens"? It also already exists, rather than producing a new semantics and a new equivalence...

Reply:
The CEK machine _is_ what really happens, but it's probably going to be harder for you to establish a connection between the translation and the CEK machine. The CEK machine should comply with some sort of small-step/big-step semantics and I'd guess that establishing the connections this way:

```
CEK <~> operational semantics <~> the translation relation
```

Also note that the CEK machine is an evaluator and not a normalizer, it's not going to look under unapplied lambdas while optimizations certainly will.

Anyway, if you can make the CEK machine work, then that's great, I just feel like you might be complicating your life.

Reply:
Yes, James and I have discussed creating a cleaner Reduction semantics but then we need to show that the CEK machine implements that, which is an extra thing to do. However, possibly a worthwhile on if there are a lot of these phases to prove.